src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs

```diff
@@ -0,0 +1,207 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem
+{
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    [Export(typeof(IBuildUpToDateCheckProvider))]
+    internal class BuildUpToDateCheck : IBuildUpToDateCheckProvider
+    {
+        private static string[] KnownOutputGroups = { "Symbols", "Built", "ContentFiles", "Documentation", "LocalizedResourceDlls" };
+
+        private readonly IProjectLockService _projectLockService;
+        private readonly IProjectSystemOptions _projectSystemOptions;
+        private readonly ConfiguredProject _configuredProject;
+        private readonly IProjectLogger _projectLogger;
+        private readonly ProjectProperties _projectProperties;
+        private readonly IProjectItemProvider _projectItemProvider;
+        private readonly IProjectItemSchemaService _projectItemsSchema;
+        private readonly Lazy<IFileTimestampCache> _fileTimestampCache;
+
+        [ImportingConstructor]
+        public BuildUpToDateCheck(
+            IProjectLockService projectLockService,
+            IProjectSystemOptions projectSystemOptions,
+            ConfiguredProject configuredProject,
+            IProjectLogger projectLogger,
+            ProjectProperties projectProperties,
+            [Import(ExportContractNames.ProjectItemProviders.SourceFiles)] IProjectItemProvider projectItemProvider,
+            [Import(AllowDefault = true)] IProjectItemSchemaService projectItemSchema,
+            [Import(AllowDefault = true)] Lazy<IFileTimestampCache> fileTimestampCache)
+        {
+            _projectLockService = projectLockService;
+            _projectSystemOptions = projectSystemOptions;
+            _configuredProject = configuredProject;
+            _projectLogger = projectLogger;
+            _projectProperties = projectProperties;
+            _projectItemProvider = projectItemProvider;
+            _projectItemsSchema = projectItemSchema;
+            _fileTimestampCache = fileTimestampCache;
+        }
+
+        private void Log(string traceMessage) => _projectLogger.WriteLine($"FastUpToDate: {traceMessage}");
+
+        private DateTime GetLatestTimestamp(List<string> paths, IDictionary<string, DateTime> timestampCache)
+        {
+            var latestTime = DateTime.MinValue;
+
+            foreach (var path in paths)
+            {
+                if (!timestampCache.TryGetValue(path, out var time))
+                {
+                    if (File.Exists(path))
+                    {
+                        time = File.GetLastWriteTimeUtc(path);
+                        timestampCache[path] = time;
+                        Log($"Path '{path}' had live timestamp '{time}'.");
+                    }
+                    else
+                    {
+                        Log($"Path '{path}' did not exist.");
```

Comment:
Yes, I thought about that over the weekend. If an input or output file is missing, we should fail the up to date check.